torch/utils/flop_counter.py

```
@@ -0,0 +1,306 @@
+) -> int:
+    """
+    Count flops for convolution. Note only multiplication is
+    counted. Computation for addition and bias is ignored.
+    Flops for a transposed convolution are calculated as
+    flops = (x_shape[2:] * prod(w_shape) * batch_size).
+    Args:
+        x_shape (list(int)): The input shape before convolution.
+        w_shape (list(int)): The filter shape.
+        out_shape (list(int)): The output shape after convolution.
+        transposed (bool): is the convolution transposed
+    Returns:
+        int: the number of flops
+    """
+    batch_size = x_shape[0]
+    conv_shape = (x_shape if transposed else out_shape)[2:]
+    c_out, c_in, *dims = w_shape
+
+    # NB(chilli): I don't think this properly accounts for padding :think:
+    # NB(chilli): Should be 2 * c_in - 1 technically for FLOPs.
+    flop = batch_size * prod(conv_shape) * c_out * prod(dims) * 2 * c_in
+    return flop
+
+def conv_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
+    """
+    Count flops for convolution.
+    """
+    x_shape, w_shape = input_shapes[:2]
+    out_shape = output_shapes[0]
+    transposed = input_shapes[6]
+
+    return conv_flop_count(x_shape, w_shape, out_shape, transposed=transposed)
+
+def transpose_shape(shape):
+    return [shape[1], shape[0]] + list(shape[2:])
+
+def conv_backward_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
+    grad_out_shape, x_shape, w_shape = input_shapes[:3]
+    output_mask = input_shapes[-1]
+    fwd_transposed = input_shapes[7]
+    flop_count = 0
+
+    if output_mask[0]:
+        grad_input_shape = get_shape(output_shapes[0])
+        flop_count += conv_flop_count(grad_out_shape, w_shape, grad_input_shape, not fwd_transposed)
+    if output_mask[1]:
+        grad_weight_shape = get_shape(output_shapes[1])
+        flop_count += conv_flop_count(transpose_shape(x_shape), grad_out_shape, grad_weight_shape, fwd_transposed)
+
+    return flop_count
+
+
+flop_mapping = {
```

Can we also add support to scaled_dot_product_attention since it's more and more used nowadays?

---

Good idea.

---

And einsum

---

Einsum isn’t needed since it’s a “compositeimplicit” op and decomposes into other operators.

---

I'm gonna do the attention ones in a follow-up PR.

---

This actually leads to an important design issue: I think there should be an API to return "unsupported" ops. Similar to https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/fvcore/nn/jit_analysis.py#L98

The reason is that users are mainly interacting with high-level modules and ops, and aren't aware of what low-level ops are called. Imagine one day torch adds a new op for transformer, or adds a different implementation of einsum that doesn't decompose: the change should be transparent, but it makes the flop counter become a silent error!

So in fvcore we provide unsupported ops and print a warning about them by default. To make the results more meaningful, we have a list of trivial ops that are always ignored and will not appear in "unsupported ops" (https://github.com/facebookresearch/fvcore/blob/51092b5515cbb493f73de079743dd6b11cc4bbf1/fvcore/nn/jit_analysis.py#L28).

---

@ppwwyyxx Hmm... I think I might just add all the ops as "unsupported ops". 

I think one benefit of putting it in core is that in principle, we should hopefully be able to keep the operator list up to date better.

---

Yeah, impl changes should be caught by tests, we have an einsum test in this PR, but not matmul?